From b82303e7023524d5dabe0d7632d4213bd0583a5a Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Wed, 14 May 2014 12:22:02 -0400
Subject: [PATCH 038/102] sparsify: Prevent overwriting block or char output
 devices (RHBZ#1056290).

virt-sparsify doesn't work if the output is a block device, and cannot
possibly work if the output is a char device.  Currently if you try
this it actually overwrites (deletes) the output device which is not
exactly desirable.  Therefore throw an error and exit.
(cherry picked from commit 13bca32e4ec9213e69ea55e837dea846719fb67e)
---
 sparsify/sparsify.ml | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/sparsify/sparsify.ml b/sparsify/sparsify.ml
index a32ef83..f6c404e 100644
--- a/sparsify/sparsify.ml
+++ b/sparsify/sparsify.ml
@@ -150,6 +150,15 @@ read the man page virt-sparsify(1).
   if contains_colon outdisk then
     error (f_"output filename '%s' contains a colon (':'); qemu-img command line syntax prevents us from using such an image") outdisk;
 
+  (* Check the output is not a block or char special (RHBZ#1056290). *)
+  if is_block_device outdisk then
+    error (f_"output '%s' cannot be a block device, it must be a regular file")
+      outdisk;
+
+  if is_char_device outdisk then
+    error (f_"output '%s' cannot be a character device, it must be a regular file")
+      outdisk;
+
   indisk, outdisk, compress, convert,
     debug_gc, format, ignores, machine_readable,
     option, quiet, verbose, trace, zeroes
-- 
1.9.3

